File: trunk/BodagheeParticleTool/BodagheeParticleTool/Emitter.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Drawing;
using BodagheeParticleTool;
using Microsoft.DirectX;
using Microsoft.DirectX.Direct3D;
using System.Threading;

namespace ParticleTool
{
    class Emitter
    {
        public List<Particle> Particle_List = new List<Particle>();
        public bool continuous;
        public bool reAnimate;
        public int image;
        public string imageFile;

        Color colorStart;
        Color colorEnd;
        private float fSize;
        private float maxLife;
        private float minVelX;
        private float maxVelX;
        private float minVelY;
        private float maxVelY;
        private float spwnAreaWidth;
        private float spwnAreaHeight;
        private float scaleStart;
        private float scaleEnd;
        private float rotation;
        private int source;
        private int dest;
        private float emitterPosX;
        private float emitterPosY;
        private int maxParticles;

        public Color ColorStart { get { return colorStart; } set { colorStart = value; } }
        public Color ColorEnd { get { return colorEnd; } set { colorEnd = value; } }
        public float FSize { get { return fSize; } set { fSize = value; } }
        public float MaxLife { get { return maxLife; } set { maxLife = value; } }
        public float MinVelX { get { return minVelX; } set { minVelX = value; } }
        public float MaxVelX {get { return maxVelX; } set { maxVelX = value; } }
        public float MinVelY { get { return minVelY; } set { minVelY = value; } }
        public float MaxVelY { get { return maxVelY; } set { maxVelY = value; } }
        public float SpwnAreaWidth {get { return spwnAreaWidth; }set { spwnAreaWidth = value; }}
        public float SpwnAreaHeight{ get { return spwnAreaHeight; }set { spwnAreaHeight = value; }}
        public float ScaleStart{ get { return scaleStart; } set { scaleStart = value; } }
        public float ScaleEnd{ get { return scaleEnd; }set { scaleEnd = value; }}
        public float Rotation{ get { return rotation; } set { rotation = value; } }
        public int Source{get { return source; }set { source = value; } }
        public int Dest{get { return dest; } set { dest = value; } }
        public float EmitterPosX{get { return emitterPosX; }set { emitterPosX = value; }}
        public float EmitterPosY{ get { return emitterPosY; }set { emitterPosY = value; }}
        public int MaxParticles { get { return maxParticles; }set { maxParticles = value; } }

        public Emitter()
        {
            image = -1;
            colorStart = Color.White;
            colorEnd = Color.White;
            fSize = 1.0f;
            maxLife = 1.0f;
            minVelX = 0;
            maxVelX = 0;
            minVelY = 0;
            maxVelY = 0;
            spwnAreaWidth = 1;
            spwnAreaHeight = 1;
            scaleStart = 0.0f;
            scaleEnd = 0.0f;
            rotation = 0.0f;
            source = 2;
            dest = 3;
            emitterPosX = 0;
            emitterPosY = 0;
            maxParticles = 0;

            continuous = true;
            reAnimate = false;
        }

        public void Update(float deltaTime)
        {
                for (int i = 0; i < maxParticles; i++)
                {
                    Particle_List[i].rotation += rotation * deltaTime;
                    Particle_List[i].life -= 1000 * deltaTime;
                    Particle_List[i].Color = ColorOperator.Lerp(ColorEnd, ColorStart, Particle_List[i].life / 1000);

                    Particle_List[i].posX += Particle_List[i].velX * deltaTime;
                    Particle_List[i].posY += Particle_List[i].velY * deltaTime;

                    if (scaleStart <= scaleEnd)
                    {
                        Particle_List[i].size_p += 0.5f * deltaTime;
                        if (Particle_List[i].size_p >= scaleEnd)
                        {
                            Particle_List[i].size_p += 0;
                        }
                    }
                    if (scaleStart >= scaleEnd)
                    {
                        Particle_List[i].size_p -= 0.5f * deltaTime;
                        if (Particle_List[i].size_p <= scaleStart)
                        {
                            Particle_List[i].size_p += 0;
                        }
                    }

                    if (Particle_List[i].Life < 0 && continuous)
                    {
                        Respawn(Particle_List[i]);
                    }
                    if ( Particle_List[i].life < 0 && reAnimate && continuous != true)
                    {
                        Respawn(Particle_List[i]);

                        if (i >= maxParticles - 1)
                        {
                            reAnimate = false;
                        }
                    }
                
            }
        }
        public void Respawn(Particle p)
        {
            Thread.Sleep(1);
            Random rand = new Random((int)DateTime.Now.Ticks & 0x0000FFFF);
            p.Life = (float)(rand.Next(0, (int)MaxLife * 1000));
            p.Size_p = FSize;

            p.VelX = rand.Next((int)(MinVelX), (int)(MaxVelX));
            p.VelY = rand.Next((int)(MinVelY), (int)(MaxVelY));

            p.posX = rand.Next( (int)emitterPosX, ( (int)(emitterPosX) + (int)(spwnAreaWidth) ) );
            p.posY = rand.Next( (int)emitterPosY, ( (int)(emitterPosY) + (int)(spwnAreaHeight) ) );
        }
        public void Render()
        {
            int tempBlend1 = 0;
            int tempBlend2 = 0;

            tempBlend1 = (int)SGP.ManagedDirect3D.Instance.Device.RenderState.SourceBlend;
            tempBlend2 = (int)SGP.ManagedDirect3D.Instance.Device.RenderState.DestinationBlend;

            SGP.ManagedDirect3D.Instance.Sprite.Flush();

            SGP.ManagedDirect3D.Instance.Device.SetRenderState(Microsoft.DirectX.Direct3D.RenderStates.SourceBlend, this.Source);
            SGP.ManagedDirect3D.Instance.Device.SetRenderState(Microsoft.DirectX.Direct3D.RenderStates.DestinationBlend, this.Dest);

            for (int i = 0; i < this.Particle_List.Count; i++)
            {
                if (this.Particle_List[i].Life > 0)
                {
                    int OffSetX = (int)((int)(SGP.ManagedTextureManager.Instance.GetTextureWidth(image) >> 1) * this.Particle_List[i].Size_p);
                    int OffSetY = (int)((int)(SGP.ManagedTextureManager.Instance.GetTextureHeight(image) >> 1) * this.Particle_List[i].Size_p);

                    SGP.ManagedTextureManager.Instance.Draw(image, (int)this.Particle_List[i].PosX,
                        (int)this.Particle_List[i].PosY, this.Particle_List[i].Size_p, this.Particle_List[i].Size_p, Rectangle.Empty, OffSetX, OffSetY,
                        this.Particle_List[i].rotation, this.Particle_List[i].Color.ToArgb());
                }
            }
            SGP.ManagedDirect3D.Instance.Sprite.Flush();
            tempBlend1 = (int)SGP.ManagedDirect3D.Instance.Device.RenderState.SourceBlend;
            tempBlend2 = (int)SGP.ManagedDirect3D.Instance.Device.RenderState.DestinationBlend;
        }
    }
}

File: trunk/Aeona Sprint 2/Aeona Project/source/Particle Engine/Emitter.cpp
#include "StdAfx.h"
#include "Emitter.h"
#include "..\CGame.h"
#include "..\Wrappers\SGD_Math.h"



Emitter::Emitter()
{
	Image = -1;
	ColorStartA = 255;
	ColorStartR = 255;
	ColorStartG = 255;
	ColorStartB = 255;
	ColorEndA   = 255;
	ColorEndR   = 255;
	ColorEndG   = 255;
	ColorEndB   = 255;
	fSize = 0.0f;
	MaxParticles = 0;
	MaxLife = 0.0f;
	MinVelX = 0.0f;
	MaxVelX = 0.0f;
	MinVelY = 0.0f;
	MaxVelY = 0.0f;
	SpawnWidth = 0.0f;
	SpawnHeight = 0.0f;
	ScaleStart = 0.0f;
	ScaleEnd = 0.0f;
	Rotation = 0.0f;
	Source = 0;
	Destination = 0;
	EmitterPosX = 0.0f;
	EmitterPosY = 0.0f;

	Continuous = true;
	reAnimate = false;

	GravityPosX = 0.0f;
	GravityPosY = 0.0f;
	GravityPower = 0.0f;
	GravityDistX = 0.0f;
	GravityDistY = 0.0f;
}
Emitter::Emitter(const Emitter& E)
{
	for(unsigned int i = 0; i < E.Particle_List.size(); i++)
		Particle_List.push_back(E.Particle_List[i]);
}
Emitter& Emitter::operator=(const Emitter& E)
{
	if(this == &E)
		return *this;
	else
		*this = E;

	return *this;
}
Emitter::~Emitter()
{
	ShutDown();
}

void Emitter::Init()
{
	Particle* tempParticle;

	for (int i = 0; i < MaxParticles; i++)
	{
		tempParticle = new Particle();
		tempParticle->Life = -1;
		Particle_List.push_back(tempParticle);
	}
}
void Emitter::Update( float DT )
{
	for( int i = 0; i < MaxParticles; ++i )
	{
		Particle_List[i]->Rotation += Rotation * DT;
		Particle_List[i]->Life -= DT;

		Particle_List[i]->A = (int)(Lerp( (float)ColorStartA, (float)ColorEndA, ( Particle_List[i]->curLife - Particle_List[i]->Life ) / Particle_List[i]->curLife ));
		Particle_List[i]->R = (int)(Lerp( (float)ColorStartR, (float)ColorEndR, ( Particle_List[i]->curLife - Particle_List[i]->Life ) / Particle_List[i]->curLife )); 
		Particle_List[i]->G = (int)(Lerp( (float)ColorStartG, (float)ColorEndG, ( Particle_List[i]->curLife - Particle_List[i]->Life ) / Particle_List[i]->curLife ));
		Particle_List[i]->B = (int)(Lerp( (float)ColorStartB, (float)ColorEndB, ( Particle_List[i]->curLife - Particle_List[i]->Life ) / Particle_List[i]->curLife ));
		fSize = Lerp( ScaleStart, ScaleEnd, ( Particle_List[i]->curLife - Particle_List[i]->Life ) / Particle_List[i]->curLife );
		Particle_List[i]->Size_p = fSize;

		GravityDistX = GravityPosX - Particle_List[i]->PosX;
		GravityDistY = GravityPosY - Particle_List[i]->PosY;

		float length;
		length = sqrt( ( GravityDistX * GravityDistX ) + ( GravityDistY * GravityDistY ) );

		GravityDistX = GravityDistX / length;
		GravityDistY = GravityDistY / length;

		GravityDistX *= GravityPower;
		GravityDistY *= GravityPower;

		Particle_List[i]->PosX += Particle_List[i]->VelX * DT;
		Particle_List[i]->PosY += Particle_List[i]->VelY * DT;

		Particle_List[i]->VelX += GravityDistX;
		Particle_List[i]->VelY += GravityDistY;

		if( Particle_List[i]->Life < 0 && Continuous )
		{
			ReSpwn( Particle_List[i] );
		}
		if( Particle_List[i]->Life < 0 && reAnimate == true && Continuous != true )
		{
			ReSpwn( Particle_List[i] );

			if( i >= MaxParticles - 1 )
			{
				reAnimate = false;
			}
		}
	}
}
void Emitter::ReSpwn(Particle* P)
{
	P->Life = RandomFloat( 0.0f, MaxLife );
	P->curLife = P->Life;
	P->Size_p = ScaleStart;

	P->Rotation = Rotation;

	P->VelX = RandomFloat( MinVelX , MaxVelX );
	P->VelY = RandomFloat( MinVelY , MaxVelY );

	P->PosX = RandomFloat( EmitterPosX, ( EmitterPosX + SpawnWidth ) );
	P->PosY = RandomFloat( EmitterPosY, ( EmitterPosY + SpawnHeight) );
}
void Emitter::Render()
{
	DWORD tempBlends[2] = {0};
	CSGD_Direct3D::GetInstance()->GetDirect3DDevice()->GetRenderState(D3DRS_SRCBLEND, &tempBlends[0]);
	CSGD_Direct3D::GetInstance()->GetDirect3DDevice()->GetRenderState(D3DRS_DESTBLEND, &tempBlends[1]);

	CSGD_Direct3D::GetInstance()->GetSprite()->Flush();

	CSGD_Direct3D::GetInstance()->GetDirect3DDevice()->SetRenderState(D3DRS_SRCBLEND, Source);
	CSGD_Direct3D::GetInstance()->GetDirect3DDevice()->SetRenderState(D3DRS_DESTBLEND, Destination);

	for (unsigned int i = 0; i < Particle_List.size(); i++ )
	{
		if ( Particle_List[i]->Life > 0)
		{
			if( Image != -1 )
			{
				int OffSetX = (int)((int)(CSGD_TextureManager::GetInstance()->GetTextureWidth( Image ) >> 1 ) * Particle_List[i]->Size_p );
				int OffSetY = (int)((int)(CSGD_TextureManager::GetInstance()->GetTextureHeight( Image ) >> 1 ) * Particle_List[i]->Size_p );

				CSGD_TextureManager::GetInstance()->Draw( Image, (int)Particle_List[i]->PosX,
					(int)Particle_List[i]->PosY, Particle_List[i]->Size_p, Particle_List[i]->Size_p, NULL, (float)OffSetX, (float)OffSetY,
					Particle_List[i]->Rotation, 
					D3DCOLOR_ARGB( Particle_List[i]->A, Particle_List[i]->R, Particle_List[i]->G, Particle_List[i]->B ) );
			}
		}
	}
	CSGD_Direct3D::GetInstance()->GetSprite()->Flush();
	CSGD_Direct3D::GetInstance()->GetDirect3DDevice()->SetRenderState(D3DRS_SRCBLEND, tempBlends[0]);
	CSGD_Direct3D::GetInstance()->GetDirect3DDevice()->SetRenderState(D3DRS_DESTBLEND, tempBlends[1]);
}
void Emitter::ShutDown(void)
{
	for(unsigned int i = 0; i < Particle_List.size(); i++)
		delete (Particle*)(Particle_List[i]);

	Particle_List.clear();

	MaxParticles = 0;
}